src/coreclr/interpreter/compiler.cpp

```diff
@@ -807,6 +859,114 @@ void InterpCompiler::EmitCode()
     m_compHnd->setBoundaries(m_methodInfo->ftn, m_ILToNativeMapSize, m_pILToNativeMap);
 }
 
+void InterpCompiler::BuildGCInfo(InterpMethod *pInterpMethod)
+{
+#ifdef FEATURE_INTERPRETER
+    InterpIAllocator* pAllocator = new (this) InterpIAllocator(this);
+    InterpreterGcInfoEncoder* gcInfoEncoder = new (this) InterpreterGcInfoEncoder(m_compHnd, m_methodInfo, pAllocator, Interp_NOMEM);
+    assert(gcInfoEncoder);
+
+    gcInfoEncoder->SetCodeLength(ConvertOffset(m_methodCodeSize));
+
+    uint32_t stackSlotCount = m_totalVarsStackSize / INTERP_STACK_SLOT_SIZE,
+        slotTableSize = stackSlotCount * sizeof(GcSlotId);
+    // [pObjects, pByrefs]
+    GcSlotId *slotTables[2] = {
+        (GcSlotId *)alloca(slotTableSize),
+        (GcSlotId *)alloca(slotTableSize)
+    };
+    // 0 is a valid slot id so default-initialize all the slots to 0xFFFFFFFF
+    for (int i = 0; i < 2; i++)
+        memset(slotTables[i], 0xFF, slotTableSize);
+
+    INTERP_DUMP("Allocating gcinfo slots for %u vars\n", m_varsSize);
+
+    // Request slot IDs for all our vars, not just IL vars
+    for (int i = 0; i < m_varsSize; i++)
+    {
+        GcSlotId *slotTable;
+        InterpVar *pVar = &m_pVars[i];
+        GcSlotFlags flags;
+        switch (pVar->interpType) {
+            case InterpTypeO:
+                slotTable = slotTables[0];
+                flags = (GcSlotFlags)0;
+                break;
+            case InterpTypeByRef:
+                slotTable = slotTables[1];
+                flags = (GcSlotFlags)GC_SLOT_INTERIOR;
+                break;
+            default:
+                continue;
+        }
+
+        if (pVar->global)
+            flags = (GcSlotFlags)(flags | GC_SLOT_UNTRACKED);
+
+        uint32_t slotIndex = pVar->offset / INTERP_STACK_SLOT_SIZE;
+        uint8_t allocateNewSlot = slotTable[slotIndex] == ((GcSlotId)-1);
+        if (allocateNewSlot)
+        {
+            // Important to pass GC_FRAMEREG_REL, the default is broken due to GET_CALLER_SP being unimplemented
+            slotTable[slotIndex] = gcInfoEncoder->GetStackSlotId(pVar->offset, flags, GC_FRAMEREG_REL);
+        }
+        else
+        {
+            assert(!pVar->global);
+        }
+
+        INTERP_DUMP(
+            "%s gcinfo slot %u for %s%svar #%d at offset %d\n",
+            allocateNewSlot ? "Allocated" : "Reused",
+            slotTable[slotIndex],
+            pVar->global ? "global " : "",
+            pVar->interpType == InterpTypeByRef ? "byref " : "",
+            i, pVar->offset
+        );
+    }
+
+    gcInfoEncoder->FinalizeSlotIds();
+
+    // Now that slot IDs are finalized we want to record live ranges for every var that isn't global
+    for (int i = 0; i < m_varsSize; i++)
+    {
+        InterpVar *pVar = &m_pVars[i];
+        // Even if we have a gc slot for this offset, this var might not be an object reference
+        if ((pVar->interpType != InterpTypeO) && (pVar->interpType != InterpTypeByRef))
```

Comment:
A change that will suck, is that here you will also need to handle valuetypes containing refs.

Reply:
I'll start thinking about how to do that. We're probably going to have to generate some sort of intermediate table and use that for both this and figuring out where/when to zero references